File: python/openjiuwen/state_provider.py
"""Python 状态契约，对应 Rust `StateProvider`。

子类实现 `snapshot(key) -> dict | StateView` 与 `report(feedback)`，经 PyO3
反向包装成 Rust trait 对象。状态是 hint：可持有跨请求记忆，超时应返回空视图。

可选扩展：`query(key, query) -> dict | StateSnapshot`。它与 `snapshot` **平级**，
两者互不影响：

- 只实现 `snapshot`：不改任何代码即可继续工作；宿主若携带检索入参，Router 会
  自动降级为 `snapshot`（`retrieved` 为空）。
- 实现 `query`：可基于 `query.text` / `query.vector` / `query.top_k` 做 KNN 等
  更精细的检索，返回 `StateSnapshot(view=..., retrieved=[RetrievedItem, ...])`。
  检索失败或返回越界时同样降级为 `snapshot`，不阻塞路由。
"""

from __future__ import annotations


class StateProvider:
    """状态槽插件契约。与算法侧 `AlgorithmProvider` 对位。"""

    name = "unnamed"

    def snapshot(self, key):
        raise NotImplementedError("{0} must implement snapshot()".format(type(self).__name__))

    def report(self, feedback):
        raise NotImplementedError("{0} must implement report()".format(type(self).__name__))

    # 可选：默认不提供。未覆写时 Router 走 snapshot 降级路径。
    # def query(self, key, query):
    #     raise NotImplementedError

File: crates/state/src/state_provider.rs
//! [`StateProvider`]：状态实现者的唯一接入点。
//!
//! 与算法侧 `AlgorithmProvider` 对位：运行期单槽选一。
//! 状态是 hint：有界、可丢失；算法从不直接调用本 trait。

use openjiuwen_protocol::{
    Feedback, RoutingKey, StateQuery, StateQueryError, StateSnapshot, StateView,
};

/// CAS 写回冲突：期望版本与当前 active 不一致。
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct CasConflict {
    pub slot: String,
    pub expected: u64,
    pub actual: u64,
}

/// 跨请求记忆的插件契约。状态是 hint：有界、可丢失。
///
/// 远程实现超时必须返回空 [`StateView`] 而非 `Err`，绝不阻塞请求。
pub trait StateProvider: Send + Sync {
    /// 路由前一次性快照。基础读取路径，入参只有路由键。
    fn snapshot(&self, key: &RoutingKey) -> StateView;

    /// 按 [`StateQuery`] 做一次检索读。与 [`StateProvider::snapshot`] 平级，不是它的带参变体。
    ///
    /// 返回 [`StateSnapshot`]：`view` 仍是按键的 hint；`retrieved` 是命中列表，
    /// `StateQuery::top_k` 只表示条数上限（实际可更少，空列表合法）。
    /// `StateQuery::route_id` 由 runtime 注入，实现可用它把本次检索与之后的
    /// [`StateProvider::report`] 对上。
    ///
    /// 默认实现忽略检索入参，只包一层 [`StateProvider::snapshot`]，既有实现
    /// 不覆盖也能工作；需要文本 / 向量检索时再覆盖。
    ///
    /// 检索失败不得阻断请求：返回仅含视图的结果，或返回 `Err` 让 runtime
    /// 回退到 `snapshot`。
    fn query(&self, key: &RoutingKey, query: &StateQuery) -> Result<StateSnapshot, StateQueryError> {
        let _ = query;
        Ok(StateSnapshot::from_view(self.snapshot(key)))
    }

    /// 吸收路由后反馈。异步、尽力而为；本骨架同步写入。
    fn report(&self, feedback: Feedback);

    /// 带版本原子写回（训练任务 → state）。默认实现为 no-op。
    fn publish(&self, slot: &str, artifact: &[u8], ver: u64) -> Result<(), CasConflict> {
        let _ = (slot, artifact, ver);
        Ok(())
    }
}
